Lib/multiprocessing/shared_memory.py

```diff
@@ -115,57 +113,146 @@ def __init__(self, name, flags=None, mode=384, size=0, read_only=False):
+_SHM_SAFE_NAME_LENGTH = 14
+
+# shared object name prefix
+if os.name == "nt":
+    _SHM_NAME_PREFIX = 'wnsm_'
+else:
+    _SHM_NAME_PREFIX = '/psm_'
+
+
+def _make_filename():
+    """Create a random filename for the shared memory object.
+    """
+    # number of random bytes to use for name
+    nbytes = (_SHM_SAFE_NAME_LENGTH - len(_SHM_NAME_PREFIX)) // 2
+    assert nbytes >= 2, '_SHM_NAME_PREFIX too long'
+    name = _SHM_NAME_PREFIX + secrets.token_hex(nbytes)
+    assert len(name) <= _SHM_SAFE_NAME_LENGTH
+    return name
+
+
+class PosixSharedMemory:
+
+    # defaults so close() and unlink() can run without errors
+    fd = -1
+    name = None
+    _mmap = None
+    _buf = None
 
     def __init__(self, name, flags=None, mode=384, size=0, read_only=False):
         if name and (flags is None):
-            _PosixSharedMemory.__init__(self, name, mode=mode)
+            flags = 0
         else:
-            if name is None:
-                name = f'psm_{os.getpid()}_{random.randrange(100000)}'
             flags = O_CREX if flags is None else flags
-            _PosixSharedMemory.__init__(
-                self,
-                name,
-                flags=flags,
-                mode=mode,
-                size=size,
-                read_only=read_only
-            )
-
+        if flags & O_EXCL and not flags & O_CREAT:
+            raise ValueError("O_EXCL must be combined with O_CREAT")
+        if name is None and not flags & O_EXCL:
+            raise ValueError("'name' can only be None if O_EXCL is set")
+        flags |= os.O_RDONLY if read_only else os.O_RDWR
+        self.flags = flags
+        self.mode = mode
+        if not size >= 0:
+            raise ValueError("'size' must be a positive integer")
+        if name is None:
+            self._open_retry()
+        else:
+            self.name = name
+            self.fd = _posixshmem.shm_open(name, flags, mode=mode)
+        if size:
+            try:
+                os.ftruncate(self.fd, size)
+            except OSError:
+                self.unlink()
+                raise
         self._mmap = mmap.mmap(self.fd, self.size)
-        self.buf = memoryview(self._mmap)
+        self._buf = memoryview(self._mmap)
+
+    @property
+    def size(self):
+        "Size in bytes."
+        if self.fd >= 0:
+            return os.fstat(self.fd).st_size
+        else:
+            return 0
+
+    @property
+    def buf(self):
+        "A memoryview of contents of the shared memory block."
+        return self._buf
+
+    def _open_retry(self):
+        # generate a random name, open, retry if it exists
+        while True:
+            name = _make_filename()
+            try:
+                self.fd = _posixshmem.shm_open(name, self.flags,
+                                               mode=self.mode)
+            except FileExistsError:
+                continue
+            self.name = name
+            break
```

Comment:
Not sure if it's worth it, just mentioning. `tempfile` module provides a max number of retries strategy you may want to reuse here:
https://github.com/python/cpython/blob/825aab95fde959541859383f8ea7e7854ebfd49f/Lib/tempfile.py#L254